src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,335 @@
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(&key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), key)))
+        }
+        if !map_descriptor.value_type.admits(&value) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.value_type), value)))
+        }
+
+        let exists = self.fetch_entry(contract_name, map_name, &key)? != Value::Void;
+        if exists {
+            return Ok(Value::Bool(false))
+        }
+
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &Some(value.serialize()?)];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
+            &params);
+
+        return Ok(Value::Bool(true))
+    }
+
+    pub fn delete_entry(&mut self, contract_name: &str, map_name: &str, key: &Value) -> Result<Value> {
+        let exists = self.fetch_entry(contract_name, map_name, &key)? != Value::Void;
+        if !exists {
+            return Ok(Value::Bool(false))
+        }
+
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), (*key).clone())))
+        }
+
+        let none: Option<String> = None;
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &none];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
+            &params);
+
+        return Ok(Value::Bool(exists))
+    }
+
+
+    pub fn get_contract(&mut self, contract_name: &str) -> Result<Contract> {
+        self.load_contract(contract_name)?
+            .ok_or_else(|| { Error::new(ErrType::UndefinedContract(contract_name.to_string())) })
+    }
+
+    pub fn insert_contract(&mut self, contract_name: &str, contract: Contract) -> Result<()> {
+        if self.load_contract(contract_name)?.is_some() {
+            Err(Error::new(ErrType::ContractAlreadyExists(contract_name.to_string())))
+        } else {
+            self.execute("INSERT INTO contracts (contract_name, contract_data) VALUES (?, ?)",
+                         &[contract_name, &contract.serialize()?]);
+            Ok(())
+        }
+    }
+
+    pub fn get_simmed_block_height(&self) -> Result<i128> {
+        let block_height: (i128) =
+            self.query_row(
+                "SELECT simmed_block_height FROM simmed_block_table LIMIT 1",
+                NO_PARAMS,
+                |row| row.get(0))
+            .expect("Failed to fetch simulated block height");
+
+        Ok(block_height)
+    }
+
+    pub fn set_simmed_block_height(&mut self, block_height: i128) {
```

Comment:
Oh interesting, I didn't know about the `i128_blob` feature.  I'd been using `u64` for block heights, and then further converting them to and from `i64`.